hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -949,16 +935,81 @@ private TcpClientConnection onAuthenticated(TcpClientConnection connection,
         return connection;
     }
 
-    private void checkClientStateOnClusterIdChange(TcpClientConnection connection) {
+    /**
+     * Checks the client state against the intend of the callee(switchingToNextCluster)
+     * closes the connection and throws exception if the authentication needs to be cancelled.
+     */
+    private void checkClientState(TcpClientConnection connection, boolean switchingToNextCluster) {
+        if (clientState == ClientState.SWITCHING_CLUSTER && !switchingToNextCluster) {
+            String reason = "There is a cluster switch in progress. "
+                    + "This connection attempt initiated before the progress and not allowed to be authenticated.";
+            connection.close(reason, null);
+            throw new AuthenticationException(reason);
+        }
+        //Following state can not happen. There is only one path with `switchingToNextCluster` as true
+        //and that path starts only when the old switch fails. There are no concurrent run of that path.
+        if (!(clientState == ClientState.SWITCHING_CLUSTER || clientState == ClientState.INITIAL)
```

Comment:
One more possible scenario:

We have cluster configs for A and B. A is up, B is down. We are connected to A.

Something bad happens, the client loses connection to A, but let's assume that A is still up.

We will try to connect A, fail, try to connect B, fail. And, imagine that the network problem between A and the client is gone. We can see it now. We will try to connect it. Since we kicked in the failover logic, `switchingToNextCluster` will be true and the state would be `INITIALIZED_ON_CLUSTER` we will close the connection again.

In this scenario, possibly we will connect back with the `ConnectToAllClusterMembersTask`. 

I am wondering if we can come up with scenarios where there is no cluster-id change, and `ConnectToAllClusterMembersTask` cannot handle that.